### PR TITLE
#0: hash build on harvesting mask

### PR DESCRIPTION
**Pull request by aliuTT**

### CODE REVIEW

**Comment by davorchap** [comment]
the key change is hash kernel binary `harvesting_mask` instead of `device_id`  ?

**Comment by davorchap** [comment]
@tt-asaigal probably good to take a look -- 8x devices will be expected to have hits in the binary cache if harvesting_mask is a match. 
Check if there's any races of 8x workers going to the common kernel binary cache.

File: tt_metal/impl/kernels/kernel.cpp [hunk 1]
```diff
@@ -114,13 +114,13 @@ uint8_t ComputeKernel::expected_num_binaries() const {
...
     int expected_num_binaries = this->expected_num_binaries();
-    if (this->binaries_.find(device_id) != this->binaries_.end() and this->binaries_.at(device_id).size() != expected_num_binaries) {
+    if (this->binaries_.find(harvesting_mask) != this->binaries_.end() and this->binaries_.at(harvesting_mask).size() != expected_num_binaries) {
         TT_THROW("Expected " + std::to_string(expected_num_binaries) + " binaries but have "
-                    + std::to_string(this->binaries_.at(device_id).size()) + " for kernel " + this->name());
+                    + std::to_string(this->binaries_.at(harvesting_mask).size()) + " for kernel " + this->name());
     }
-    return this->binaries_.at(device_id);
+    return this->binaries_.at(harvesting_mask);
 }
```
Comment (tt-asaigal) [inline tt_metal/impl/kernels/kernel.cpp:124]: Maybe I'm missing something, but kernels belong to a program and programs don't span devices. Why do we have a device_id/harvesting_mask map here?

File: tt_metal/jit_build/build.hpp [hunk 2]
```diff
@@ -165,11 +165,11 @@ void jit_build(const JitBuildState& build, const JitBuildSettings *settings, con
...
 void jit_build_subset(const JitBuildStateSubset& builds, const JitBuildSettings *settings, const string& kernel_in_path);
 
-inline const string jit_build_get_kernel_compile_outpath(int device_id) {
+inline const string jit_build_get_kernel_compile_outpath(int harvesting_mask) {
     // TODO(pgk), get rid of this
     // The test infra needs the output dir.  Could put this in the device, but we plan
     // to remove the device dependence in the future, so putting this here for now
-    return llrt::OptionsG.get_root_dir() + "/built/" + std::to_string(device_id) + "/kernels/";
+    return llrt::OptionsG.get_root_dir() + "/built/" + std::to_string(harvesting_mask) + "/kernels/";
 }
```
[the remark this answers is not in the document]
Comment (pgkeller) [reply tt_metal/jit_build/build.hpp:173]: Yes, need a lock here.  Create dir will be racy and may fail (which is ok), then create/acquire the lock.  The loser has to spin and wait until the lock is released and then should be ready to go.

File: tt_metal/impl/kernels/kernel.cpp (hunk 1 above)
Comment (pgkeller) [reply tt_metal/impl/kernels/kernel.cpp:124]: kernels can be compiled for multiple devices, so one kernel can have multiple instances, need to grab the right one.  and, this is a dependency that we need to break (fix that is) - should only have one instance of a kernel that is device agnostic

**Review by pgkeller — CHANGES_REQUESTED** [review]

**Review by ttmtrajkovic — APPROVED** [review]

File: tt_metal/third_party/tt_llk_grayskull
Comment (rtawfik01) [inline tt_metal/third_party/tt_llk_grayskull:1]: Please run `git submodule update --init --recursive`. The commit has older submodule links added

File: tt_metal/third_party/tt_llk_wormhole_b0
Comment (rtawfik01) [inline tt_metal/third_party/tt_llk_wormhole_b0:1]: Please run `git submodule update --init --recursive`. The commit has older submodule links added

**Review by rtawfik01 — CHANGES_REQUESTED** [review]

**Review by rtawfik01 — APPROVED** [review]

**Review by pgkeller — APPROVED** [review]

**Pull request merged**